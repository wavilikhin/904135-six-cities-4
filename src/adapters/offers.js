export const createOffer = (offer) => ({
  id: offer.id,
  city: offer.city.name,
  cityZoom: offer.city.location.zoom,
  isPremium: offer.is_premium,
  cityCoords: [offer.city.location.latitude, offer.city.location.longitude],
  image: offer.preview_image,
  priceValue: offer.price,
  name: offer.title,
  type: offer.type,
  coords: [offer.location.latitude, offer.location.longitude],
  bedrooms: offer.bedrooms,
  description: offer.description,
  goods: offer.goods,
  host: {
    avatar_url: offer.host.avatar_url,
    id: offer.host.id,
    isPro: offer.host.is_pro,
    name: offer.host.name,
  },
  images: offer.images,
  isFavorite: offer.is_favorite,
  location: {
    latitude: offer.location.latitude,
    longitude: offer.location.longitude,
    zoom: offer.location.zoom,
  },
  maxAdults: offer.max_adults,
  rating: offer.rating,
});
