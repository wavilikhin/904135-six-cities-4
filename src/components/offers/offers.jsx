import React from 'react';

import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import OffersList from '../offers-list/offers-list.jsx';
import Map from '../map/map.jsx';

const Offers = (props) => {
  const { offersDataArray, city } = props;

  const filtredOffers = offersDataArray.filter((offer) => {
    return offer.city === city;
  });

  return (
    <div className="cities">
      <div className="cities__places-container container">
        <section className="cities__places places">
          <h2 className="visually-hidden">Places</h2>
          <b className="places__found">
            {filtredOffers.length} places to stay in {city}
          </b>
          <form className="places__sorting" action="#" method="get">
            <span className="places__sorting-caption">Sort by</span>
            <span className="places__sorting-type" tabIndex="0">
              Popular
              <svg className="places__sorting-arrow" width="7" height="4">
                <use xlinkHref="#icon-arrow-select"></use>
              </svg>
            </span>
            <ul className="places__options places__options--custom places__options--opened">
              <li
                className="places__option places__option--active"
                tabIndex="0"
              >
                Popular
              </li>
              <li className="places__option" tabIndex="0">
                Price: low to high
              </li>
              <li className="places__option" tabIndex="0">
                Price: high to low
              </li>
              <li className="places__option" tabIndex="0">
                Top rated first
              </li>
            </ul>
          </form>
          <OffersList />
        </section>
        <div className="cities__right-section">
          <section className="cities__map map">
            <Map city={city} zoom={12} offers={filtredOffers} />
          </section>
        </div>
      </div>
    </div>
  );
};

Offers.propTypes = {
  offersDataArray: PropTypes.arrayOf(
    PropTypes.shape({
      city: PropTypes.string.isRequired,
      cityCoords: PropTypes.arrayOf(PropTypes.number, PropTypes.number),
      quality: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
      priceValue: PropTypes.string.isRequired,
      priceText: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      coords: PropTypes.arrayOf(PropTypes.number, PropTypes.number).isRequired,
    }),
  ).isRequired,
  city: PropTypes.string,
};

const mapStateToProps = (state) => ({
  city: state.city,
  offersDataArray: state.offers,
});

export { Offers };
export default connect(mapStateToProps, null)(Offers);
