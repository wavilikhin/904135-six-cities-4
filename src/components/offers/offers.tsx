import * as React from 'react';
import { connect } from 'react-redux';
import OffersList from '../offers-list/offers-list';
import Map from '../map/map';
import { getCity, getFiltredOffers } from '../../reducer/data/selectors';
import MainEmpty from '../main-empty/main-empty';
import PlaceSorting from '../place-sorting/place-sorting';
import { OfferInfo } from '../../types';

interface Props {
  filtredOffers: OfferInfo[];
  city: string;
}

const Offers: React.FC<Props> = ({ filtredOffers, city }) => {
  return (
    <div className="cities">
      <div className="cities__places-container container">
        <section className="cities__places places">
          <h2 className="visually-hidden">Places</h2>
          <b className="places__found">
            {filtredOffers.length} places to stay in {city}
          </b>
          <PlaceSorting />
          {filtredOffers.length === 0 ? (
            <MainEmpty city={city} />
          ) : (
            <OffersList />
          )}
        </section>
        <div className="cities__right-section">
          <section className="cities__map map">
            <Map offers={filtredOffers} />
          </section>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => ({
  city: getCity(state),
  filtredOffers: getFiltredOffers(state),
});

export { Offers };
export default connect(mapStateToProps, null)(Offers);
