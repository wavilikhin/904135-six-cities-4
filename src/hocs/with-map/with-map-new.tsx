import * as React from 'react';
import * as leaflet from 'leaflet';
import { connect } from 'react-redux';
import { OfferInfo } from '../../types';
import { AppStateType } from '../../reducer/reducer';
import { getHoveredOffer } from '../../reducer/state/selectors';
import { Diff } from 'utility-types';

type StateToPropsTypes = {
  hoveredOffer: OfferInfo;
};

type OwnProps = {
  offers: OfferInfo[];
};

type InjectedPropsTypes = {};

type HocProps = StateToPropsTypes & OwnProps;

type State = {
  offers: OfferInfo[] | null;
  map: leaflet.Map | null;
  mapLayer: leaflet.Layer | null;
  markersLayer: leaflet.Marker[] | null;
};

export const withMap = <BasePropsTypes extends InjectedPropsTypes>(
  Component: React.ComponentType<BasePropsTypes>,
) => {
  const mapStateToProps = (state: AppStateType): StateToPropsTypes => ({
    hoveredOffer: getHoveredOffer(state),
  });

  class WithMap extends React.PureComponent<HocProps, State> {
    props: HocProps;
    state: State;

    constructor(props) {
      super(props);

      this.state = {
        map: null,
        mapLayer: null,
        offers: null,
        markersLayer: null,
      };
    }

    componentDidMount() {
      const { offers } = this.props;
      // Creating base map
      let map = leaflet.map(`map`, {
        center: [51.22172, 6.1],
        zoom: 12,
        zoomControl: false,
      });
      // Setting base map layer
      const mapLayer = leaflet
        .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
          {
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
          },
        )

        .addTo(map);
      // If the only provided offer === mocked offer returning base map
      if (offers.length === 1 && offers[0].city === '') {
        map.setView([51.22172, 6.77616], 12);

        this.setState({
          map,
          mapLayer,
        });
        return;
      }
      // If offers are provided setting up markers layer
      const cityLat: number = offers[0].cityCoords[0];
      const cityLon: number = offers[0].cityCoords[1];
      const cityCoordinates: leaflet.LatLngTuple = [cityLat, cityLon];

      const zoom: number = offers[0].cityZoom;

      map.setView(cityCoordinates, zoom);

      let markersLayer: leaflet.Marker[] = [];

      offers.forEach((offer) => {
        const icon: leaflet.Icon = leaflet.icon({
          iconUrl: `/img/pin.svg`,
          iconSize: [30, 30],
          className: `offer-${offer.id}`,
        });

        markersLayer.push(
          leaflet.marker([offer.location.latitude, offer.location.longitude], {
            icon,
          }),
        );
      });

      let offersLayer = leaflet.layerGroup(markersLayer);

      offersLayer.addTo(map);

      this.setState({
        map,
        mapLayer,
        offers,
      });
    }

    componentDidUpdate() {
      console.count('updated');
      const { offers, hoveredOffer } = this.props;

      // Handling pin change for hovered offer
      const hoveredOfferId = hoveredOffer?.id;

      // If hoveredOffer = undefined making all pins basic color
      if (!hoveredOfferId && this.state.markersLayer) {
        let currentMarkers = this.state.markersLayer;
        currentMarkers.forEach((marker) =>
          marker.setIcon(
            leaflet.icon({
              iconUrl: `/img/pin.svg`,
              iconSize: [30, 30],
              className: `offer-${
                marker.options.icon.options.className.split('-')[1]
              }`,
            }),
          ),
        );
        // If hoveredOffer != undefined making it's pin active color
      } else if (hoveredOffer && this.state.markersLayer) {
        let currentMarkers = this.state.markersLayer;
        if (currentMarkers !== null) {
          currentMarkers.forEach((marker) => {
            if (
              Number(marker.options.icon.options.className.split('-')[1]) ===
              Number(hoveredOfferId)
            ) {
              marker.setIcon(
                leaflet.icon({
                  iconUrl: `/img/pin-active.svg`,
                  iconSize: [30, 30],
                  className: `offer-${
                    marker.options.icon.options.className.split('-')[1]
                  }`,
                }),
              );
            }
          });
        }
      }

      if (offers.length === 0) return;

      const cityLat: number = offers[0].cityCoords[0];
      const cityLon: number = offers[0].cityCoords[1];
      const cityCoordinates: leaflet.LatLngTuple = [cityLat, cityLon];

      const zoom: number = offers[0].cityZoom;

      let map = this.state.map;
      let mapLayer = this.state.mapLayer;
      const currentOffers = this.state.offers;

      map.setView(cityCoordinates, zoom);

      // If offers array changed updating pins layer
      if (currentOffers !== offers) {
        map.eachLayer(function (layer: leaflet.LayerGroup): void {
          if (layer !== mapLayer) {
            map.removeLayer(layer);
          }
        });

        let markersLayer: leaflet.Marker[] = [];

        offers.forEach((offer) => {
          let icon = leaflet.icon({
            iconUrl: `/img/pin.svg`,
            iconSize: [30, 30],
            className: `offer-${offer.id}`,
          });
          markersLayer.push(
            leaflet.marker(
              [offer.location.latitude, offer.location.longitude],
              {
                icon,
              },
            ),
          );
        });

        let offersLayer = leaflet.layerGroup(markersLayer);
        offersLayer.addTo(map);

        this.setState({
          map,
          offers,
          markersLayer,
        });
      }

      this.setState({
        map,
      });
    }

    render() {
      const { hoveredOffer, offers, ...propsToPass } = this.props;
      return <Component {...(propsToPass as BasePropsTypes)} />;
    }
  }

  const ConnectedHoc = connect<
    StateToPropsTypes,
    {},
    Diff<BasePropsTypes, InjectedPropsTypes>,
    AppStateType
  >(
    mapStateToProps,
    null,
  )(WithMap);

  return ConnectedHoc;
};
