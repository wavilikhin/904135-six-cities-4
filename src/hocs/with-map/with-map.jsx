import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';

export const withMap = (Component) => {
  class WithMap extends PureComponent {
    constructor(props) {
      super(props);
    }

    componentDidMount() {
      const { offers } = this.props;

      const icon = leaflet.icon({
        iconUrl: `img/pin.svg`,
        iconSize: [30, 30],
      });
      this.icon = icon;

      let cityCoords;
      let zoom;

      offers.length > 0
        ? (cityCoords = offers[0].cityCoords)
        : (cityCoords = [0, 0]);

      offers.length > 0 ? (zoom = offers[0].cityZoom) : (zoom = 12);

      const map = leaflet.map(`map`, {
        center: cityCoords,
        zoom: zoom,
        zoomControl: false,
        marker: true,
      });

      map.setView(cityCoords, zoom);

      leaflet
        .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
          {
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
          },
        )
        .addTo(map);

      let markersLayer = [];

      for (const offer in offers) {
        markersLayer.push(offer);
      }

      let offersLayer = leaflet.layerGroup(markersLayer);
      offersLayer.addTo(map);

      this.offersLayer = offersLayer;

      this.map = map;
    }

    componentDidUpdate() {
      const { offers } = this.props;

      let updatedCoords;
      let updatedZoom;

      offers.length > 0
        ? (updatedCoords = offers[0].cityCoords)
        : (updatedCoords = [0, 0]);

      offers.length > 0
        ? (updatedZoom = offers[0].cityZoom)
        : (updatedZoom = 12);

      this.map.setView(updatedCoords, updatedZoom);

      let markersLayer = [];

      offers.forEach((offer) => {
        markersLayer.push(leaflet.marker(offer.coords, this.icon));
      });

      this.offersLayer.remove();

      let offersLayer = leaflet.layerGroup(markersLayer);

      this.offersLayer = offersLayer;

      offersLayer.addTo(this.map);
    }

    render() {
      return <Component />;
    }
  }

  WithMap.propTypes = {
    offers: PropTypes.arrayOf(
      PropTypes.shape({
        city: PropTypes.string.isRequired,
        cityZoom: PropTypes.number.isRequired,
        isPremium: PropTypes.bool.isRequired,
        image: PropTypes.string.isRequired,
        priceValue: PropTypes.string.isRequired,
        priceText: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        coords: PropTypes.arrayOf(PropTypes.number).isRequired,
      }),
    ).isRequired,
  };
  return WithMap;
};
