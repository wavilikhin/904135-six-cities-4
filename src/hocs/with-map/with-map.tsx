import React, { PureComponent } from "react";
import leaflet from "leaflet";
import { OfferInfo } from "../../components/offer-card/offer-card";
interface Props {
  offers: OfferInfo[];
}

export const withMap = (Component) => {
  class WithMap extends PureComponent<Props> {
    props: Props;

    constructor(props) {
      super(props);
    }

    componentDidMount() {
      const { offers } = this.props;

      const icon = leaflet.icon({
        iconUrl: `img/pin.svg`,
        iconSize: [30, 30],
      });
      this.icon = icon;

      if (offers.length === 0) return;

      let cityCoords = offers[0].cityCoords;
      let zoom = offers[0].cityZoom;

      const map = leaflet.map(`map`, {
        center: cityCoords,
        zoom: zoom,
        zoomControl: false,
        marker: true,
      });

      map.setView(cityCoords, zoom);

      leaflet
        .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
          {
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
          }
        )
        .addTo(map);

      let markersLayer = [];

      for (const offer of offers) {
        markersLayer.push(
          leaflet.marker(
            [offer.location.latitude, offer.location.longitude],
            icon
          )
        );
      }

      let offersLayer = leaflet.layerGroup(markersLayer);
      offersLayer.addTo(map);

      this.offersLayer = offersLayer;

      this.map = map;
    }

    componentDidUpdate() {
      const { offers } = this.props;
      console.log(offers);

      if (offers.length === 0) return;

      let updatedCoords = offers[0].cityCoords;
      let updatedZoom = 12;

      this.map
        ? this.map.setView(updatedCoords, updatedZoom)
        : (this.map = leaflet.map(`map`, {
            center: updatedCoords,
            zoom: updatedZoom,
            zoomControl: false,
            marker: true,
          }));

      // this.map
      //   ? (updatedMap = this.map)
      //   : (updatedMap = leaflet.map(`map`, {
      //       center: updatedCoords,
      //       zoom: updatedZoom,
      //       zoomControl: false,
      //       marker: true,
      //     }));

      this.map.setView(updatedCoords, updatedZoom);
      leaflet
        .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
          {
            attribution:
              '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
          }
        )
        .addTo(this.map);

      let markersLayer = [];
      offers.forEach((offer) => {
        markersLayer.push(
          leaflet.marker(
            [offer.location.latitude, offer.location.longitude],
            this.icon
          )
        );
      });

      this.offersLayer ? this.offersLayer.remove() : null;

      let offersLayer = leaflet.layerGroup(markersLayer);
      offersLayer.addTo(this.map);

      this.offersLayer = offersLayer;
    }

    render() {
      return <Component />;
    }
  }

  return WithMap;
};
