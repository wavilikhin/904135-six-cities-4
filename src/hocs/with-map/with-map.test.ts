// main
import * as React from 'react';
import renderer from 'react-test-renderer';
// components
import { withMap } from './with-map.jsx';
// data
import { OFFERS } from '../../test/__mocks__/offers.js';
const mockData = OFFERS;

const MockComponent = () => {
  return <div style={{ height: `100vh` }} id="map"></div>;
};

const MockComponentWrapped = withMap(MockComponent);

it(`WithMap snapthot test`, () => {
  const tree = renderer
    .create(<MockComponentWrapped offers={mockData} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
