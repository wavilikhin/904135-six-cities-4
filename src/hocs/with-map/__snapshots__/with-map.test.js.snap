// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithMap snapthot test 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "100vh",
    }
  }
/>
`;
